### DISCUSSION TITLE
SIWE requires 2 user actions?

### DISCUSSION THREAD

**Discussion by e-roy** [wevm/wagmi#200] [Q&A] [answered]
So from what I understand, using SIWE takes 2 separate user actions, one to connect wallet, another to sign message.

From the developer point of view, I would like this to be one button on the App's UI for the user to press to make both these actions (it would create 2 separate wallet user interactions of course)

I tried to create this flow with a promise, but half the time I get an error.  After the wallet is connected, and trying to invoke a function to sign message I will get this error about 50% of the time.

```
ConnectorNotFoundError: Connector not found
    at wagmi.esm.js:1095:31
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (wagmi.esm.js:50:20)
    at _next (wagmi.esm.js:72:9)
    at wagmi.esm.js:79:7
    at new Promise (<anonymous>)
    at wagmi.esm.js:68:12
    at wagmi.esm.js:1133:19
```

The user could press the button again to be able to sign message and SIWE will work fine.

Is there something I can do to reduce the frequency of this error, hopefully to none? (I have slipped a setTimeout function inside my promise, but can't tell how much if it does reduce the error)

**Comment by tmm** [comment, marked as answer]
Added a [component](https://github.com/tmm/wagmi/blob/main/docs/components/web3/WalletSelectorWithSiwe.tsx) to the docs site that connects wallet and signs the SIWE message in one step. After using the selected `connector` to connect, you use it to sign the SIWE message:

```tsx
try {
  const res = await connect(connector) // connect from useConnect
  if (!res.data) throw res.error ?? new Error('Something went wrong')

  const nonceRes = await fetch('/api/nonce')
  const message = new SiweMessage({
    domain: window.location.host,
    address: res.data.account,
    statement: 'Sign in with Ethereum to the app.',
    uri: window.location.origin,
    version: '1',
    chainId: res.data.chain?.id,
    nonce: await nonceRes.text(),
  })

  const signer = await connector.getSigner()
  const signature = await signer.signMessage(message.prepareMessage())

  const verifyRes = await fetch('/api/verify', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ message, signature }),
  })
  if (!verifyRes.ok) throw new Error('Error verifying message')

  // It worked! User is signed in with Ethereum
} catch (error) {
  // Do something with the error
}
```

Will consider making this the default in the docs, but having two distinct steps is a little bit easier to grok IMO.

**Reply by tmm** [reply]
Added to guide too

https://wagmi-xyz.vercel.app/guides/sign-in-with-ethereum#combine-connect-wallet-and-sign--verify-message-steps

**Reply by Kipitup** [reply]
Hello @tmm,

I've tried to implement this, but `await connect(connector)` does not return anything here.

The main documentation suggest using [onConnect](https://wagmi.sh/docs/hooks/useConnect#onconnect-optional) callback function when configuring `useConnect`. This solution lack flexibility. You can't access data from other wagmi hook and putting a big function in the `useConnect` config makes it hard to read.

I tried to compartmentalize it: 

```javascript
const { connect, connectors, error } = useConnect({ onConnect })
```

```javascript
const onConnect = async (data: any) => {

  try {
    const nonceRes = await fetch('/api/nonce')
    const message = new SiweMessage({
      domain: window.location.host,
      address: data.account,
      statement: 'Sign in with Ethereum to the app.',
      uri: window.location.origin,
      version: '1',
      chainId: data.chain?.id,
      nonce: await nonceRes.text(),
    })

    const signer = await data.connector.getSigner()
    const signature = await signer.signMessage(message.prepareMessage())

    console.log('signature :', signature)

    const verifyRes = await fetch('/api/verify', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ message, signature }),
    })

    if (!verifyRes.ok)
      throw new Error('Error verifying message')

  } catch (error) {
    console.error(error);
  }
}

export { onConnect }
```

Questions:

- What is the correct type for `data` here ?
- How am I supposed to do error management ? In your SIWE example, you use a React state to do this. While it's technically possible to send a React setter function all the way to `onConnect`, this seems to be the wrong method.

Thanks for your help

**Reply by tmm** [reply]
Try `connectAsync` instead of `connect`. That part of the docs needs to be updated to use it as well.

**Reply by Kipitup** [reply]
Awesome, thanks a lot. I submitted a PR for the SIWE example. https://github.com/tmm/wagmi/pull/476
